File: src/nova_audio_agent/realtime/memory_board.py
"""Read-only Memory Board serialization: one bounded frame per explicit request.

The board is a viewer, not a second dialogue truth: it renders the live process's
`memory.channels` snapshot and persists nothing (R100).
"""

from __future__ import annotations

import json
from typing import Any

MAX_BOARD_MESSAGE_BYTES = 256 * 1024
MAX_BOARD_ITEMS_PER_CHANNEL = 50
MAX_BOARD_CONTENT_CHARS = 2048
MAX_BOARD_SUMMARY_CHARS = 4096


def memory_board_message(request_id: str, memory: Any) -> str:
    """Serialize every channel's newest items into one bounded text frame.

    When the encoded frame exceeds the byte bound, the oldest retained items are
    dropped round-robin across channels; if fixed fields (summaries) still exceed
    the bound, summaries are dropped entirely rather than returning an over-budget
    frame.
    """
    channels = [_channel_view(channel) for channel in memory.channels.values()]
    summaries_dropped = False
    while True:
        message = json.dumps(
            {"type": "memory.board", "request_id": request_id, "channels": channels},
            ensure_ascii=False,
            separators=(",", ":"),
        )
        if len(message.encode("utf-8")) <= MAX_BOARD_MESSAGE_BYTES:
            return message
        populated = [channel for channel in channels if channel["items"]]
        if populated:
            largest = max(populated, key=lambda channel: len(channel["items"]))
            largest["items"] = largest["items"][1:]
            continue
        if not summaries_dropped:
            summaries_dropped = True
            for channel in channels:
                channel["summary"] = None
            continue
        return message


def _channel_view(channel: Any) -> dict[str, Any]:
    items = channel.items[-MAX_BOARD_ITEMS_PER_CHANNEL:]
    summary = channel.summary
    if summary is not None:
        summary = summary[:MAX_BOARD_SUMMARY_CHARS]
    return {
        "name": channel.name,
        "summary": summary,
        "uncompressed": channel.uncompressed,
        "item_count": len(channel.items),
        "items": [_item_view(item) for item in items],
    }


def _item_view(item: Any) -> dict[str, Any]:
    content = json.dumps(item.content, ensure_ascii=False, separators=(",", ":"))
    truncated = len(content) > MAX_BOARD_CONTENT_CHARS
    view: dict[str, Any] = {
        "seq": item.seq,
        "ts": item.ts,
        "trust": item.trust,
        "priority": item.priority,
        "outcome": item.outcome,
        "refs": list(item.refs),
        "content": content[:MAX_BOARD_CONTENT_CHARS],
    }
    if truncated:
        view["truncated"] = True
    return view

File: runtime/src/realtime/memory-board.ts
import { Buffer } from 'node:buffer'
import type { Channel, Memory, MemoryItem } from '../memory.js'
import type {RealtimeDiagnosticsSnapshot} from './telemetry.js'

export const MAX_BOARD_MESSAGE_BYTES = 256 * 1024
export const MAX_BOARD_ITEMS_PER_CHANNEL = 50
export const MAX_BOARD_CONTENT_CHARS = 2048
export const MAX_BOARD_SUMMARY_CHARS = 4096
export const MAX_BOARD_REFRESH_MESSAGE_BYTES = 64 * 1024
export const MAX_BOARD_REFRESH_ITEMS_PER_CHANNEL = 12
export const MAX_BOARD_REFRESH_CONTENT_CHARS = 768
export const MAX_BOARD_REFRESH_SUMMARY_CHARS = 1024
export const MAX_BOARD_REFRESH_DIAGNOSTICS = 32

export type MemoryBoardDetail = 'compact' | 'full'

export interface MemoryBoardMessageOptions {
  readonly detail?: MemoryBoardDetail
}

interface BoardItem {
  readonly seq: number
  readonly ts: number
  readonly trust: MemoryItem['trust']
  readonly priority: number
  readonly outcome: MemoryItem['outcome']
  readonly refs: readonly string[]
  readonly content: string
  readonly truncated?: true
}

interface BoardChannel {
  readonly name: string
  summary: string | null
  readonly uncompressed: number
  readonly item_count: number
  items: BoardItem[]
}

export function memoryBoardMessage(
  requestId: string,
  memory: Memory,
  diagnosticSnapshot: RealtimeDiagnosticsSnapshot = {version: 1, records: []},
  options: MemoryBoardMessageOptions = {},
): string {
  const detail = options.detail ?? 'full'
  const profile = detail === 'compact'
    ? {
        messageBytes: MAX_BOARD_REFRESH_MESSAGE_BYTES,
        itemsPerChannel: MAX_BOARD_REFRESH_ITEMS_PER_CHANNEL,
        contentChars: MAX_BOARD_REFRESH_CONTENT_CHARS,
        summaryChars: MAX_BOARD_REFRESH_SUMMARY_CHARS,
        diagnostics: MAX_BOARD_REFRESH_DIAGNOSTICS,
      }
    : {
        messageBytes: MAX_BOARD_MESSAGE_BYTES,
        itemsPerChannel: MAX_BOARD_ITEMS_PER_CHANNEL,
        contentChars: MAX_BOARD_CONTENT_CHARS,
        summaryChars: MAX_BOARD_SUMMARY_CHARS,
        diagnostics: diagnosticSnapshot.records.length,
      }
  const channels = [...memory.channels.values()].map(channel => channelView(channel, profile))
  const diagnostics = {
    version: 1 as const,
    records: diagnosticSnapshot.records.slice(-profile.diagnostics),
  }
  let summariesDropped = false
  while (true) {
    const message = JSON.stringify({
      type: 'memory.board', request_id: requestId, diagnostics, channels,
    })
    if (Buffer.byteLength(message, 'utf8') <= profile.messageBytes) return message
    if (diagnostics.records.length > 0) {
      diagnostics.records = diagnostics.records.slice(1)
      continue
    }
    const populated = channels.filter(channel => channel.items.length > 0)
    if (populated.length > 0) {
      const largest = populated.reduce((current, candidate) => (
        candidate.items.length > current.items.length ? candidate : current
      ))
      largest.items = largest.items.slice(1)
      continue
    }
    if (!summariesDropped) {
      summariesDropped = true
      for (const channel of channels) channel.summary = null
      continue
    }
    return message
  }
}

interface BoardProfile {
  readonly itemsPerChannel: number
  readonly contentChars: number
  readonly summaryChars: number
}

function channelView(channel: Channel, profile: BoardProfile): BoardChannel {
  return {
    name: channel.name,
    summary: channel.summary === null
      ? null
      : sliceCodePoints(channel.summary, profile.summaryChars),
    uncompressed: channel.uncompressed,
    item_count: channel.items.length,
    items: channel.items
      .slice(-profile.itemsPerChannel)
      .map(item => itemView(item, profile.contentChars)),
  }
}

function itemView(item: MemoryItem, contentChars: number): BoardItem {
  const content = JSON.stringify(item.content)
  const truncated = [...content].length > contentChars
  return {
    seq: item.seq,
    ts: item.ts,
    trust: item.trust,
    priority: item.priority,
    outcome: item.outcome,
    refs: [...item.refs],
    content: sliceCodePoints(content, contentChars),
    ...(truncated ? {truncated: true as const} : {}),
  }
}

function sliceCodePoints(value: string, limit: number): string {
  return [...value].slice(0, limit).join('')
}
